### PR TITLE
Nodejs express - optimize copy for buildah

### PR DESCRIPTION
**Pull request by neeraj-laad**
### Checklist:

We are seeing significant performance issues with all Node.js based stacks to build the final application image when using buildah (instead of docker).

This is because of a known issue with buildah 1.9.0. We can not upgrade the buildah version because of other issues.

This is a temporary fix to optimize the number of files we copy during the build process.

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stacks-overview.md#stack-structure).

### Modifying an existing stack:

- [x] Updated the stack version in `stack.yaml`

### CODE REVIEW

File: incubator/nodejs-express/image/project/Dockerfile [hunk 1]
```diff
@@ -28,13 +29,19 @@ RUN apt-get update \
  && echo 'Finished installing dependencies'
 
 # Copy the project
-COPY --chown=node:node . /project
```
Comment (sam-github) [inline incubator/nodejs-express/image/project/Dockerfile]: Does buildah do better  when copying between images than it does when copying from host?

It looks like all the lines from here to `ENV NODE_PATH` could be replaced by a single line:

> COPY --chown=node:node --from=0 /project /project

Comment (EnriqueL8) [reply incubator/nodejs-express/image/project/Dockerfile]: In the first docker image we don't have all the files from the `/project` dir, since we only copy the `package.json` for the project and the user-app. I agree that we could simplify the process by first copying the tars, and `package.json'`s with that one command and then copy the files from the host.

File: incubator/nodejs/image/project/Dockerfile [hunk 2]
```diff
@@ -7,11 +7,12 @@ RUN apt-get update \
  && apt-get clean \
  && echo 'Finished installing dependencies'
 
+WORKDIR "/project/user-app/node_modules"
```
Comment (EnriqueL8) [inline incubator/nodejs/image/project/Dockerfile]: Are the `nodejs` specific changes in the other PR also?

File: incubator/nodejs/stack.yaml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 name: Node.js
```
Comment (EnriqueL8) [inline incubator/nodejs/stack.yaml]: Also this change, should be remove since we have another PR for nodejs

File: incubator/nodejs/image/project/Dockerfile (hunk 2 above)
Comment (sam-github) [reply incubator/nodejs/image/project/Dockerfile]: Yes: https://github.com/appsody/stacks/pull/679#issuecomment-589350889

File: incubator/nodejs-express/image/project/Dockerfile [hunk 4]
```diff
@@ -8,15 +8,16 @@ RUN apt-get update \
  && echo 'Finished installing dependencies'
 
 # Install stack dependencies
+WORKDIR /project/node_modules
```
[the remark this answers is not in the document]
Comment (neeraj-laad) [reply incubator/nodejs-express/image/project/Dockerfile]: The intent was to always create the node_modules folder as if there are no dependencies then it might not get created and then the tar commands could fail. But in this case, this is not really an issue as we know exactly what is in the stack. 

Deleting this line.

File: incubator/nodejs-express/image/project/Dockerfile [hunk 5]
```diff
@@ -8,15 +8,16 @@ RUN apt-get update \
...
 WORKDIR /project
 COPY ./package*.json ./
-RUN npm install --production
+RUN npm install --production && tar czf stack_modules.tgz node_modules && rm -rf node_modules
 
 # Install user-app dependencies
+WORKDIR /project/user-app/node_modules
```
[the remark this answers is not in the document]
Comment (neeraj-laad) [reply incubator/nodejs-express/image/project/Dockerfile]: Wil change to `RUN mkdir -p <path>` to be clear about what we are trying to do here.

File: incubator/nodejs-express/image/project/Dockerfile (hunk 1 above)
Comment (neeraj-laad) [reply incubator/nodejs-express/image/project/Dockerfile]: The issue we have today is that buildah 1.9.0 does not honour `.dockerignore`, and has a buggy algorithm for iterating over files. We can not move to a higher version (which has this fix) yet because of other issues.

I can move the copy commands in the first stage, however, we will still need those separate copy statements in the first stage then. Would we just move the problem around to the first stage?

**Comment by neeraj-laad** [comment]
@sam-github Thanks for the review comments. I accidentally created the PR with changes to multiple stacks. It is easier to release stacks individually so have split the changes in 2 PRs again.

I have added comments to provide the link to the PR that has further links to original issue. I hope that helps provide some context to the issue.

**Comment by sam-github** [comment]
> The issue we have today is that buildah 1.9.0 does not honour .dockerignore, and has a buggy algorithm for iterating over files. We can not move to a higher version (which has this fix) yet because of other issues.

^--- without this information, the Dockerfile isn't understandable on its own, so its likely to get refactored to break buildah. I strongly suggest that the information above be put into the dockerfile in a comment. It doesn't necessarily need a link to this PR in the comment, since that can be found from `git blame`.

Also, I'm a bit confused by "we can't update", who is "we"? appsody doesn't use buildah, does it? Since buildah isn't used in appsody AFAICT, its not possible to know if any future change here breaks whoever is using buildah downstream (who is using buildah downstream?). At some point in the future, I assume this structure can be deleted, but without any reproduction information here, or link to the downstream consumer who is having this issue, it won't be possible to know or ask when the workaround can be deleted.

**Comment by neeraj-laad** [comment]
Appsody uses buildah instead of docker when used in a Kubernetes environment like Che hosted IDE and Tekton build pipelines. appsody/tekton-sample repo even provides sample Tekton pipelines for user's to start with.

The version of buildah that we use today 1.9.0 has this performance issue during copy. Newer version of buildah have resolved this but they have only works if newer versions of  `fuseoverlayfs` is installed on the host (not the container). Given the changes are needed on the host it makes it difficult for Apprody team to control and fix.

**Comment by neeraj-laad** [comment]
@sam-github I have re-worked this PR as we were copying node_moduels from user's app over in the final stage. This meant that dev/test dependencies fro muser's app were still getting into the final application image. 

I have now moved the copying of the application over in the first stage and I delete node_modules before running `npm i --production`. 

This has also cleaned up the copy steps a bit. 

Please let me know your thoughts.

File: incubator/nodejs-express/image/project/Dockerfile [hunk 6]
```diff
@@ -7,16 +7,23 @@ RUN apt-get update \
  && apt-get clean \
  && echo 'Finished installing dependencies'
 
+# Copying individual files/folders as buildah 1.9.0 does not honour .dockerignore
+COPY package*.json ./
```
Comment (groeges) [inline incubator/nodejs-express/image/project/Dockerfile]: @neeraj-laad Is this COPY correct? Where are the package*.json files getting copied to?
Where this was moved from we had done a `WORKDIR /project` first so I presume the .json files were being copied into /project, but with it as we have now are they getting copied into the correct directory.

Comment (neeraj-laad) [reply incubator/nodejs-express/image/project/Dockerfile]: Good catch..

**Comment by neeraj-laad** [comment]
@sam-github Would like to get a final review from you on this.
